### PR TITLE
Use FindPython3 when cmake version >= 3.15.0

### PR DESCRIPTION
**Pull request by okhowang**
Fix https://github.com/ARMmbed/mbed-crypto/issues/389

## Status
**READY**

## Requires Backporting
NO

## Migrations
NO

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Some CI jobs are failing on Ubuntu 16.04 which has cmake 3.5.1.
```
-- The C compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:53 (if):
  if given arguments:

    "CMAKE_VERSION" "VERSION_GREATER_EQUAL" "3.12.0"

  Unknown arguments specified


-- Configuring incomplete, errors occurred!
See also "/var/lib/build/CMakeFiles/CMakeOutput.log".
```
I can't retrieve `CMakeOutput.log` from the CI but hopefully this should be enough to reproduce the problem.

**Comment by okhowang** [comment]
VERSION_GREATER_EQUAL was introduced in came v3.7. It's my mistake

**Comment by okhowang** [comment]
I notice that `PYTHON_EXECUTABLE` is also used in `3rdparty/everest/CMakeLists.txt`.
Is it need to be update too?
@gilles-peskine-arm

**Comment by gilles-peskine-arm** [comment]
@okhowang Oh yes please update that one too. And the `find_package(PythonInterp)` in `tests/CMakeLists.txt` might need updating as well? I'm not even sure whether that's needed, or will the toplevel cmakelists take care of it?

jenkins-internal is an internal-website, yes. Unfortunately we can't make it public. Hopefully soon we'll transfer our CI to infrastructure which is maintained under https://trustedfirmware.org and which is publicly readable. In the meantime, please ask an Arm team member for CI results. Here the CI is passing except for a known unrelated issue under Mbed OS. But our CI doesn't use cmake ≥3.12.

**Comment by okhowang** [comment]
I have modified `PYTHON_EXECUTABLE` in `3rdparty/everest/CMakeLists.txt`.

There is a comment with `Cannot build test suites without Python 2 or 3` in `tests/CMakeLists.txt`.
which mean that test can run with python 2 or 3.
I also test it use docker image `python:2`, and it's OK for test.
I think it's no needs to modify `test/CMakeLists.txt`

**Comment by gilles-peskine-arm** [comment]
Thanks for the update!

> Cannot build test suites without Python 2 or 3

This comment is obsolete: we no longer support Python 2. Our older scripts were written to be compatible with both versions, and it's possible that we haven't broken Python 2 compatibility yet. But the build scripts should only be looking for Python 3 these days.

**Comment by gilles-peskine-arm** [comment]
With cmake 3.16.3, when I run `cmake ..` from a subdirectory, I see both
```
-- Found Python3: /usr/bin/python3.8 (found version "3.8.2") found components: Interpreter 
…
-- Found PythonInterp: /usr/bin/python3.8 (found version "3.8.2") 
```
Shouldn't it be only one or the other depending on the cmake version?

**Comment by okhowang** [comment]
now `tests/CMakeLists.txt` use `MBEDTLS_PYTHON_EXECUTABLE` too

**Comment by gilles-peskine-arm** [comment]
Unfortunately older versions of cmake are now failing (see CI results on travis).

**Comment by okhowang** [comment]
```
CMake Error at tests/CMakeLists.txt:21 (message):
  Cannot build test suites without Python 3
-- Configuring incomplete, errors occurred!
See also "/home/travis/build/ARMmbed/mbedtls/CMakeFiles/CMakeOutput.log".
See also "/home/travis/build/ARMmbed/mbedtls/CMakeFiles/CMakeError.log".
The command "cmake -D CMAKE_BUILD_TYPE:String="Check" ." exited with 1.
```
it seems like there is no Python 3 in CI environment?

**Comment by gilles-peskine-arm** [comment]
Oh. Similar builds are working on Jenkins. Locally, a cmake builds works on Ubuntu 16.04 (cmake 3.5.1), Ubuntu 18.04 (cmake 3.10.2) and Ubuntu 20.04 (cmake 3.16.3). So it does look like there's something wrong on Travis. But that can't just be a missing Python 3: `tests/scripts/check-files.py` succeeds and it's a `#!/usr/bin/env python3` script. I don't see anything about Python versions in the Travis log and I don't know where the Dockerfile is. I'll look for it.

**Comment by gilles-peskine-arm** [comment]
If I understand correctly, we're using the default environment on Travis, which is [Ubuntu 16.04 plus a few additions](https://docs.travis-ci.com/user/reference/xenial/), including Python 3.6.7 and CMake 3.12.4.

**Comment by okhowang** [comment]
I think I found key reason of failed in travis CI.
my test is in [here](https://travis-ci.org/github/okhowang/mbedtls/jobs/672410395)
according cmake v3.12.4 [FindPython/Support.cmake](https://github.com/Kitware/CMake/blob/v3.12.4/Modules/FindPython/Support.cmake#L213)
findpython3 will try a [list](https://github.com/Kitware/CMake/blob/v3.12.4/Modules/FindPython/Support.cmake#L20) of python version
and test is it available in [L268](https://github.com/Kitware/CMake/blob/v3.12.4/Modules/FindPython/Support.cmake#L268) by execute `$python -c "import sys; sys.stdout.write('.'.join([str(x) for x in sys.version_info[:3]]))"`

in travis CI. python3 has version 3.5, but there is python3.6 and python3.7 in PATH, see [here](https://travis-ci.org/github/okhowang/mbedtls/jobs/672410395#L856) which is [unavailable](https://travis-ci.org/github/okhowang/mbedtls/jobs/672410395#L903) yet.

so cmake find python3.7 first and judge python 3.7 not available, then end with python3 not found.

**Comment by okhowang** [comment]
BTW, cmake [3.15](https://cmake.org/cmake/help/latest/release/3.15.html#modules) has option `Python3_FIND_STRATEGY=LOCATION` to continue scan more python version in this case.

**Comment by okhowang** [comment]
@gilles-peskine-arm I see two internal testing failed

**Comment by gilles-peskine-arm** [comment]
Mbed OS testing failed, but don't worry about it, that's a preexisting problem. The CI results are acceptable.

File: .travis.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ sudo: false
 cache: ccache
 
 script:
+- rm -f `which python3.6` `which python3.7`
```
Comment (gilles-peskine-arm) [inline .travis.yml:9]: This is not robust. What if the environment is changed to add `python3.8`?

And what if a similar problem occurs in an environment that isn't Travis? As far as I understand, the goal is to work around a bug in CMake's `find_package` command. Ok, a workaround doesn't have to be pretty. But I'd expect the workaround to be in `CMakeLists.txt`, not in the Travis configuration.

Comment (okhowang) [reply .travis.yml]: Now, I find Python 3.5 exact in travis.
Is it OK?

Comment (gilles-peskine-arm) [reply .travis.yml]: No, that's no better. It's still based on “is this running on travis?” and not based on finding a python version that works.

Comment (okhowang) [reply .travis.yml]: > BTW, cmake 3.15 has option Python3_FIND_STRATEGY=LOCATION to continue scan more python version in this case.

there is a solution based on finding a python version that works.
but it require cmake 3.15 which is not exists on travis.

Comment (Patater) [reply .travis.yml]: Which cmake version does one get when installing cmake with apt on travis?

```diff
 addons:
   apt:
     packages:
     - doxygen
     - graphviz
+    - cmake
```

Comment (okhowang) [reply .travis.yml]: 3.12

**Comment by gilles-peskine-arm** [comment]
I'm labelling this as "needs: work" because I don't think making an exception for Travis is correct: the same problem could arise in any other environment which happens to have multiple Python versions installed in a certain way. But I'm leaving the "needs: review" label because I'm open to a third opinion that would convince me that Travis, specifically, is doing something wrong. In that case the workaround should be in `.travis.yml` like it was earlier, not in `CMakeLists.txt`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -50,18 +50,29 @@ set(CTR_DRBG_128_BIT_KEY_WARNING "${WARNING_BORDER}"
                          "${WARNING_BORDER}")
 
 # Python 3 is only needed here to check for configuration warnings.
-find_package(PythonInterp 3)
-if(PYTHONINTERP_FOUND)
+if(NOT CMAKE_VERSION VERSION_LESS 3.15.0)
```
Comment (ronald-cron-arm) [inline CMakeLists.txt:53]: Please update the commit message accordingly, 3.15.0 and not 3.12.0.

Comment (okhowang) [reply CMakeLists.txt:53]: done

**Comment by okhowang** [comment]
shall we continue?

**Comment by okhowang** [comment]
> We could do with a ChangeLog entry. Otherwise looks good.

Should I add an entry in `ChangeLog.d`?

**Pull request merged**